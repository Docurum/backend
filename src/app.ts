import cors from "cors";
import express, { Express, NextFunction, Request, Response } from "express";
import rateLimit from "express-rate-limit";
import helmet from "helmet";
import createError, { HttpError } from "http-errors";
import morgan from "morgan";
import path from "path";
// import favicon from "serve-favicon";
import xss from "xss-clean";

import "@v1/config/env.config";

import config from "@v1/config";
// import { authMiddleware } from "@v1/middlewares";
import { authRoutes, clinicRoutes, forumRoutes, userRoutes } from "@v1/routes";

// RateLimitter
const limiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 1000, // Limit each IP to 1000 requests per `window` (here, per 15 minutes)
  standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
  legacyHeaders: false, // Disable the `X-RateLimit-*` headers
  message: {
    status: createError.TooManyRequests().status,
    message: createError.TooManyRequests().message,
  },
});

const corsOption = {
  origin: [config.FRONTEND_URL, "http://localhost:3000", "https://www.docurum.com"],
};

const app: Express = express();

// Global variable appRoot with base dirname
global.appRoot = path.resolve(__dirname);

// Middlewares
app.use(helmet());
app.set("trust proxy", 1);
app.use(limiter);
app.use(xss());
app.use(cors(corsOption));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(morgan("dev"));
// app.use(favicon(path.join(__dirname, "public", "favicon.ico")));
// app.use(authMiddleware);

// Welcome Route
app.all("/", (_req: Request, res: Response, _next: NextFunction) => {
  res.send({ message: "API is Up and Running 😎🚀" });
});

const apiVersion: string = "v1";

// Routes
app.use(`/${apiVersion}/auth`, authRoutes);
app.use(`/${apiVersion}/forum`, forumRoutes);
app.use(`/${apiVersion}/user`, userRoutes);
app.use(`/${apiVersion}/clinic`, clinicRoutes);
// app.use(`/${apiVersion}/consult`, consultationRoutes);

// 404 Handler
app.use((_req: Request, _res: Response, next: NextFunction) => {
  next(createError.NotFound());
});

// Error Handler
app.use((err: HttpError, _req: Request, res: Response, _next: NextFunction) => {
  res.status(err.status || 500);
  res.send({
    status: err.status || 500,
    message: err.message,
  });
});

export default app;
