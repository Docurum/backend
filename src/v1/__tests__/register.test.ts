import app from "../../app";
import supertest from "supertest";

describe("register user", () => {
  it("should throw error for no payload", async () => {
    const resp = await supertest(app).post("/v1/auth/register");
    expect(resp.body.status).toBe(500);
    const error = JSON.stringify(resp.body, null, 2);
    expect(error).toMatchSnapshot();
  });
  it("should throw error for invalid payload", async () => {
    const userPayload = {
      name: "   abc   def   ",
      email: "abc",
      username: "_s4_.sf",
      password: "123",
      confirmPassword: "123",
      isDoctor: false,
      abc: "def",
    };
    const resp = await supertest(app).post("/v1/auth/register").send(userPayload);
    expect(resp.body.status).toBe(500);
    const error = JSON.stringify(resp.body, null, 2);
    expect(error).toMatchSnapshot();
  });
  it("should pass for correct payload", async () => {
    const userPayload = {
      name: "pinaki bhattacharjee",
      email: "pinakipb2@gmail.com",
      username: "pinakipb2",
      password: "Abcdef@$12345",
      confirmPassword: "Abcdef@$12345",
      isDoctor: false,
    };
    const resp = await supertest(app).post("/v1/auth/register").send(userPayload);
    expect(resp.status).toBe(201);
    expect(resp.body).toStrictEqual({ ...userPayload, name: "Pinaki Bhattacharjee" });
  });
});
