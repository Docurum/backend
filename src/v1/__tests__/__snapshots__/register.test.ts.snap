// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register user should throw error for invalid payload 1`] = `
"{
  "status": 500,
  "message": [
    {
      "validation": "regex",
      "code": "invalid_string",
      "message": "Invalid",
      "path": [
        "username"
      ]
    },
    {
      "validation": "email",
      "code": "invalid_string",
      "message": "Invalid email",
      "path": [
        "email"
      ]
    },
    {
      "code": "too_small",
      "minimum": 8,
      "type": "string",
      "inclusive": true,
      "exact": false,
      "message": "String must contain at least 8 character(s)",
      "path": [
        "password"
      ]
    },
    {
      "validation": "regex",
      "code": "invalid_string",
      "message": "Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character.",
      "path": [
        "password"
      ]
    },
    {
      "code": "too_small",
      "minimum": 8,
      "type": "string",
      "inclusive": true,
      "exact": false,
      "message": "String must contain at least 8 character(s)",
      "path": [
        "confirmPassword"
      ]
    },
    {
      "validation": "regex",
      "code": "invalid_string",
      "message": "Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character.",
      "path": [
        "confirmPassword"
      ]
    },
    {
      "code": "unrecognized_keys",
      "keys": [
        "abc"
      ],
      "path": [],
      "message": "Unrecognized key(s) in object: 'abc'"
    }
  ]
}"
`;

exports[`register user should throw error for no payload 1`] = `
"{
  "status": 500,
  "message": [
    {
      "code": "invalid_type",
      "expected": "string",
      "received": "undefined",
      "path": [
        "name"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "received": "undefined",
      "path": [
        "username"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "received": "undefined",
      "path": [
        "email"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "received": "undefined",
      "path": [
        "password"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "received": "undefined",
      "path": [
        "confirmPassword"
      ],
      "message": "Required"
    }
  ]
}"
`;
